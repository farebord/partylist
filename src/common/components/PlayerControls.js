import React from 'react'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { switchPlaying } from '../actions'

import { withStyles } from '@material-ui/styles';
import Grid from '@material-ui/core/Grid';
import Fab from '@material-ui/core/Fab';
import PlayIcon from '@material-ui/icons/PlayArrow';
import PauseIcon from '@material-ui/icons/Pause'
import CircularProgress from '@material-ui/core/CircularProgress';

const styles = {
    controls: {
      flexGrow: '0',
      padding: '0 10px'
    },
    songInfo: {

    },
    loadingProgress: {
        margin: '10px',
        color: 'white'
    },
    songName: {

    },
    songArtists: {

    },
    songProgress: {

    }
  }

const renderButtonIcon = (loading, isPlaying, classes) =>{
    if(loading)
        return <CircularProgress size="small" className={classes.loadingProgress} />
    if(isPlaying)
        return <PauseIcon />
    else
        return <PlayIcon />
}

const getArtists = artistList => artistList.map(artist => artist.name).join(', ')

const getTimeFromMs = ms => {
    const minutes = Math.floor(ms / 60000)
    const seconds = ((ms % 60000) / 1000).toFixed(0)
    return `${minutes}:${seconds < 10 ? '0' : ''}${seconds}`
}

export const getProgress = (progress, duration) => `${getTimeFromMs(progress)} / ${getTimeFromMs(duration)}`
  

export const PlayerControls = ({progress, songInfo, classes, isPlaying, pauseResumePlayer, loading}) =>
    <React.Fragment>
        {songInfo &&
        <Grid container>
            <Grid item className={classes.controls} xs>
                <Fab size="small" color="secondary" aria-label="Add" onClick={pauseResumePlayer}>
                    {renderButtonIcon(loading, isPlaying, classes)}
                </Fab>
            </Grid>
            <Grid container item direction="column" xs justify="center">
                <Grid item className={classes.songName}>{songInfo.name}</Grid>
                <Grid item className={classes.songArtists}>{getArtists(songInfo.artists)}</Grid>
            </Grid>
            <Grid item container justify="center" alignItems="center" className={classes.songProgress}>
                {getProgress(progress, songInfo.duration_ms)}
            </Grid>
        </Grid>
        }
    </React.Fragment>

PlayerControls.propTypes = {
    progress: PropTypes.number,
    songInfo: PropTypes.object,
    classes: PropTypes.object,
    isPlaying: PropTypes.bool,
    switchPlaying: PropTypes.func,
    loading: PropTypes.bool,
}

export const mapStateToProps = ({player}) => ({
    progress: player.progress_ms,
    songInfo: player.item,
    isPlaying: player.is_playing,
    loading: player.playbackSwitching,
})

export const mapDispatchToProps = dispatch => ({
    pauseResumePlayer: () => dispatch(switchPlaying())
})

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(PlayerControls));