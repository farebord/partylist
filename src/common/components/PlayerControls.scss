%flex-centered {
  display: flex;
  align-items: center;
  justify-content: center;
}

.playerControl {
  @extend %flex-centered;
  flex: 1 0 auto;
}

.controls {
  flex: 0 0 auto;
  padding: 0 1rem;
};

.songInfo {
  @extend %flex-centered;
  flex-direction: column;
  flex: 1 0 auto;
}

.loadingProgress {
  margin: 0.5rem;
  color: white;
};

.songProgress {
  @extend %flex-centered;
  flex: 0 0 auto;
  padding-right: 1rem;
};