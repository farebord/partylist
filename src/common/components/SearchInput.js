import React from 'react';
import { connect } from 'react-redux';
import SearchIcon from '@material-ui/icons/Search';
import Paper from '@material-ui/core/Paper';
import InputBase from '@material-ui/core/InputBase';

import './SearchInput.scss';

export const SearchInput = () => (
  <Paper className="fieldContainer">
    <InputBase
      className="searchInput"
      placeholder="Search..."
      inputProps={{ 'aria-label': 'Search...' }}
      fullWidth
    />
    <span className="inputIcon">
      <SearchIcon />
    </span>
  </Paper>
);

SearchInput.propTypes = {

};

export const mapStateToProps = () => ({

});

export const mapDispatchToProps = () => ({

});

export default connect(null, mapDispatchToProps)(SearchInput);
