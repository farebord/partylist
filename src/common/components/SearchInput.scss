%flex-centered {
  display: flex;
  align-items: center;
  justify-content: center;
}

.fieldContainer {
  @extend %flex-centered;
  padding: 0 0.25rem;
  flex: 1 0 auto;
  margin: 0 0.625rem;
}

.searchInput {
  margin-left: 0.5rem;
  flex: 1 1 auto;
}

.inputIcon {
  @extend %flex-centered;
  padding: 0.5rem;
  margin: 0;
}