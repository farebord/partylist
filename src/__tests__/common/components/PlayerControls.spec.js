import React from 'react';
import { shallow } from 'enzyme'
import { PlayerControls, getProgress } from 'common/components/PlayerControls'
import Grid from '@material-ui/core/Grid';
import PlayIcon from '@material-ui/icons/PlayArrow';
import Fab from '@material-ui/core/Fab';
import PauseIcon from '@material-ui/icons/Pause'
import CircularProgress from '@material-ui/core/CircularProgress';


describe('PlayerControls component should', () => {
    it('render correctly', () => {
        const component = shallow(<PlayerControls />)
        expect(component).toMatchSnapshot('PlayerControls renders with no props');
    });

    it('not render anything if songInfo is undefined', () => {
        const classes = {
            songName: "songName",
            songArtists: "songArtists"
        }
        const songInfo = undefined

        const component = shallow(<PlayerControls classes={classes} songInfo={songInfo} />)
        expect(component.find(Grid)).toHaveLength(0)
    })

    it('render song and artist', () => {
        const classes = {
            songName: "songName",
            songArtists: "songArtists"
        }
        const songInfo = {
            name: "TestName",
            artists: [
                {
                    name: "TestArtist"
                }
            ]
        }

        const component = shallow(<PlayerControls classes={classes} songInfo={songInfo} />)
        const songName = component.find('.songName');
        const songArtist = component.find('.songArtists')
        expect(songName.text()).toEqual('TestName')
        expect(songArtist.text()).toEqual('TestArtist')
    })

    it('render a play button if isPlaying is false', () => {
        const classes = {}

        const songInfo = {
            name: "",
            artists: []
        }

        const component = shallow(<PlayerControls songInfo={songInfo} classes={classes} loading={false} isPlaying={false} />)
        const button = component.find(PlayIcon);
        expect(button).toHaveLength(1)
    })
    
    
    it('render a pause button if isPlaying is true', () => {
        const classes = {}

        const songInfo = {
            name: "",
            artists: []
        }

        const component = shallow(<PlayerControls songInfo={songInfo} classes={classes} loading={false} isPlaying={true} />)
        const button = component.find(PauseIcon);
        expect(button).toHaveLength(1)
    })

    it('render a loading indicator in the button if loading is true', () => {
        const classes = {}

        const songInfo = {
            name: "",
            artists: []
        }

        const component = shallow(<PlayerControls songInfo={songInfo} classes={classes} loading={true} isPlaying={true} />)
        const button = component.find(CircularProgress);
        expect(button).toHaveLength(1)
    })

    it('render song duration and progress', () => {
        const classes = {
            songProgress: "songProgress"
        }

        const songInfo = {
            name: "",
            artists: [],
            duration_ms: 5000
        }

        const component = shallow(<PlayerControls progress={2000} songInfo={songInfo} classes={classes} loading={false} isPlaying={true} />)
        const progress = component.find('.songProgress');
        expect(progress.text()).toEqual('0:02 / 0:05')
    })

    it('dispatch switchPlaying on play button click', () => {
        const classes = {}

        const songInfo = {
            name: "",
            artists: []
        }
        
        const switchPlaying = jest.fn()

        const component = shallow(<PlayerControls pauseResumePlayer={switchPlaying} songInfo={songInfo} classes={classes} loading={false} isPlaying={true} />)
        component.find(Fab).simulate('click');
        expect(switchPlaying).toHaveBeenCalledTimes(1)
    })
})

describe('getProgress function should', () => {
    it('return the right progress', () => {
        const secondsProgress = 5
        const secondsDuration = 60
        const expectedReturn = "0:05 / 1:00"
        expect(getProgress(secondsProgress * 1000, secondsDuration * 1000)).toEqual(expectedReturn)
    })
})