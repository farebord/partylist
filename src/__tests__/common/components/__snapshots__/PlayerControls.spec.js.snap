// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerControls component should render correctly: PlayerControls renders with no props 1`] = `ShallowWrapper {}`;
